File: kumbaja/synthesis.py
"""Final synthesis engine — generates a structured summary of a deliberation."""

from __future__ import annotations

import json
from typing import Any

from .agent import AgentRunner
from .types import AgentConfig, DeliberationConfig, TurnRecord

SYNTHESIS_SYSTEM_PROMPT = """\
You are a deliberation synthesis agent. Your job is to read the full transcript
of a multi-agent deliberation and produce a structured summary.

Your output must be valid JSON with this exact structure:
{
  "key_arguments": ["argument 1", "argument 2", ...],
  "points_of_agreement": ["agreement 1", ...],
  "unresolved_tensions": ["tension 1", ...],
  "recommended_decision": "...",
  "confidence": "high|medium|low"
}

Be concise but thorough. Capture the essential insights from the deliberation.
"""


class SynthesisEngine:
    """Generates a final synthesis from a deliberation transcript."""

    def __init__(self, runner: AgentRunner) -> None:
        self.runner = runner

    def synthesize(
        self,
        records: list[TurnRecord],
        topic: str,
        config: DeliberationConfig,
    ) -> dict[str, Any]:
        """Run a synthesis agent to summarize the deliberation.

        Returns the parsed JSON summary.
        """
        transcript_text = self._format_transcript(records)
        envelope = {
            "topic": topic,
            "transcript": transcript_text,
            "num_agents": len(config.agents),
            "total_turns": len([r for r in records if r.agent_id != "orchestrator"]),
        }

        # Determine synthesis model
        model = config.synthesis_model or config.agents[0].model
        synth_agent = AgentConfig(
            id="synthesizer",
            model=model,
            system_prompt=SYNTHESIS_SYSTEM_PROMPT,
        )

        try:
            content, _ = self.runner.run(synth_agent, envelope)
            # Try to extract JSON from the response
            return self._extract_json(content)
        except Exception as exc:
            return {
                "key_arguments": [],
                "points_of_agreement": [],
                "unresolved_tensions": [],
                "recommended_decision": f"Synthesis failed: {exc}",
                "confidence": "low",
            }

    def _format_transcript(self, records: list[TurnRecord]) -> str:
        lines: list[str] = []
        for r in records:
            lines.append(f"[Turn {r.turn}] {r.agent_id}: {r.content}")
        return "\n".join(lines)

    def _extract_json(self, content: str) -> dict[str, Any]:
        """Extract JSON from potentially markdown-wrapped response."""
        import re

        # Look for JSON in code blocks
        code_block = re.search(r"```(?:json)?\s*(\{.*?\})\s*```", content, re.DOTALL)
        if code_block:
            content = code_block.group(1)

        # Find the first { and last }
        start = content.find("{")
        end = content.rfind("}")
        if start >= 0 and end > start:
            content = content[start : end + 1]

        return json.loads(content)

File: internal/synthesis/synthesis.go
// Package synthesis produces a structured summary from a deliberation transcript.
package synthesis

import (
	"fmt"
	"strings"

	"github.com/jgabor/agora/internal/agent"
	"github.com/jgabor/agora/internal/llmutil"
	"github.com/jgabor/agora/internal/types"
)

// DefaultSystemPrompt instructs the synthesis model to return structured JSON.
const DefaultSystemPrompt = `You are a deliberation synthesis agent. Your job is to read the full transcript
of a multi-agent deliberation and produce a structured summary.

Your output must be valid JSON with this exact structure:
{
  "key_arguments": ["argument 1", "argument 2", ...],
  "points_of_agreement": ["agreement 1", ...],
  "unresolved_tensions": ["tension 1", ...],
  "recommended_decision": "...",
  "confidence": "high|medium|low"
}

Be concise but thorough. Capture the essential insights from the deliberation.
`

// Synthesize runs a synthesis agent to summarize the deliberation.
func Synthesize(runner agent.Runner, records []types.TurnRecord, topic, model string) map[string]any {
	engine := synthesisEngine{runner: runner}
	return engine.synthesize(records, topic, model)
}

type synthesisEngine struct {
	runner agent.Runner
}

func (se *synthesisEngine) synthesize(records []types.TurnRecord, topic, model string) map[string]any {
	transcriptText := se.formatTranscript(records)

	totalTurns := 0
	for _, r := range records {
		if r.AgentID != "moderator" {
			totalTurns++
		}
	}

	envelope := map[string]any{
		"topic":       topic,
		"transcript":  transcriptText,
		"total_turns": totalTurns,
	}

	synthAgent := types.AgentConfig{
		ID:           "synthesizer",
		Model:        model,
		SystemPrompt: DefaultSystemPrompt,
	}

	content, _, err := se.runner.Run(agent.WithReadOnlyAgentPrompt(synthAgent), envelope)
	var result map[string]any
	if err != nil {
		result = map[string]any{
			"key_arguments":        []any{},
			"points_of_agreement":  []any{},
			"unresolved_tensions":  []any{},
			"recommended_decision": "Synthesis could not run: the model was unable to produce a response.",
			"confidence":           "low",
		}
	} else {
		if err := llmutil.ExtractJSON(content, &result); err != nil {
			result = map[string]any{
				"key_arguments":        []any{},
				"points_of_agreement":  []any{},
				"unresolved_tensions":  []any{},
				"recommended_decision": "Synthesis could not produce a structured summary. The model response is available in the transcript.",
				"confidence":           "low",
			}
		}
	}

	return attachCanonicalClaims(result, records)
}

func attachCanonicalClaims(result map[string]any, records []types.TurnRecord) map[string]any {
	if result == nil {
		result = make(map[string]any)
	}
	for i := len(records) - 1; i >= 0; i-- {
		if records[i].Control == nil {
			continue
		}
		claims := append([]types.ClaimEvidence(nil), records[i].Control.Claims...)
		result["claims"] = claims
		return result
	}
	// The synthesis prompt does not own claim evidence. Do not let an
	// untyped model field create claim/source output for legacy transcripts.
	delete(result, "claims")
	return result
}

func (se *synthesisEngine) formatTranscript(records []types.TurnRecord) string {
	var lines []string
	for _, r := range records {
		lines = append(lines, fmt.Sprintf("[Turn %d] %s: %s", r.Turn, r.AgentID, r.Content))
	}
	return strings.Join(lines, "\n")
}
